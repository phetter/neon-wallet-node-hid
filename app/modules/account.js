import { getAccountFromWIFKey, getPublicKeyEncoded, getAccountFromPublicKey } from 'neon-js'

import { ledgerNanoSGetPublicKey } from './ledgerNanoS'

// Constants
export const LOGIN = 'LOGIN'
export const LOGOUT = 'LOGOUT'
export const SET_DECRYPTING = 'SET_DECRYPTING'
export const SET_KEYS = 'SET_KEYS'

// Actions
export function login (wif) {
  return {
    type: LOGIN,
    wif: wif
  }
}

export function ledgerNanoSGetLogin () {
  return {
    type: LOGIN,
    ledgerNanoS: true,
    publicKey: ledgerNanoSGetPublicKey
  }
}

export function logout () {
  return {
    type: LOGOUT
  }
}

export function decrypting (bool) {
  return {
    type: SET_DECRYPTING,
    state: bool
  }
}

export function setKeys (keys) {
  return {
    type: SET_KEYS,
    keys
  }
}

// Reducer that manages account state (account now = private key)
export default (state = {wif: null, address: null, loggedIn: false, redirectUrl: null, decrypting: false, accountKeys: []}, action) => {
  switch (action.type) {
    case LOGIN:
      let loadAccount
      try {
        if (action.ledgerNanoS) {
          loadAccount = getAccountFromPublicKey(getPublicKeyEncoded(action.publicKey))
        } else {
          loadAccount = getAccountFromWIFKey(action.wif)
        }
      } catch (e) {
        process.stdout.write('error loadAccount "' + e + '" "' + e.message + '" \n')
        console.log(e.stack)
        loadAccount = -1
      }
      process.stdout.write('interim loadAccount "' + loadAccount + '" \n')
      if (loadAccount === -1 || loadAccount === -2 || loadAccount === undefined) {
        return {...state, wif: action.wif, loggedIn: false}
      }
      return {...state, wif: action.wif, address: loadAccount.address, loggedIn: true, decrypting: false}
    case LOGOUT:
      return {...state, 'wif': null, address: null, 'loggedIn': false, decrypting: false}
    case SET_DECRYPTING:
      return {...state, decrypting: action.state}
    case SET_KEYS:
      return {...state, accountKeys: action.keys}
    default:
      return state
  }
}
