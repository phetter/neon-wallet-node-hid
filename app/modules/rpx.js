// Constants
export const UPDATE_RPX_BALANCE = 'UPDATE_RPX_BALANCE'

// Actions
export function updateRpxBalance (balance) {
  return {
    type: UPDATE_RPX_BALANCE,
    RPX: balance
  }
}

// reducer for wallet account balance
export default (state = { RPX: 0 }, action) => {
  switch (action.type) {
    case UPDATE_RPX_BALANCE:
      return { ...state, 'RPX': action.RPX }
    default:
      return state
  }
}
