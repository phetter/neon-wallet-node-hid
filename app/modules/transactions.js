import { ASSETS_LABELS } from '../core/constants'
// Constants
export const SEND_TRANSACTION = 'SEND_TRANSACTION'
export const CLEAR_TRANSACTION = 'CLEAR_TRANSACTION'
export const TOGGLE_ASSET = 'TOGGLE_ASSET'

// Actions
export function sendEvent (success, message) {
  return {
    type: SEND_TRANSACTION,
    success: success,
    message: message
  }
}

export function clearTransactionEvent () {
  return {
    type: CLEAR_TRANSACTION
  }
}

export function toggleAsset () {
  return {
    type: TOGGLE_ASSET
  }
}

// Reducer for state used when performing a transaction
export default (state = { success: null, message: null, selectedAsset: ASSETS_LABELS.NEO }, action) => {
  switch (action.type) {
    case SEND_TRANSACTION:
      return {...state, success: action.success, message: action.message}
    case CLEAR_TRANSACTION:
      return {...state, success: null, message: null}
    case TOGGLE_ASSET:
      let asset
      if (state.selectedAsset === ASSETS_LABELS.NEO) {
        asset = ASSETS_LABELS.GAS
      } else {
        asset = ASSETS_LABELS.NEO
      }
      return {...state, success: null, selectedAsset: asset}
    default:
      return state
  }
}
