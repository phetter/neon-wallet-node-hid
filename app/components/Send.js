// @flow
import React, { Component } from 'react'
import { connect } from 'react-redux'
import { verifyAddress } from 'neon-js'
import { togglePane } from '../modules/dashboard'
import { sendEvent, clearTransactionEvent, toggleAsset } from '../modules/transactions'
import SplitPane from 'react-split-pane'
import ReactTooltip from 'react-tooltip'
import { log } from '../util/Logs'
import { ASSETS, ASSETS_LABELS } from '../core/constants'
import { ledgerNanoSGetdoSendAsset } from '../modules/ledgerNanoS'

type Props = {
  dispatch: DispatchType,
  address: string,
  wif: string,
  neo: number,
  net: NetworkType,
  gas: number,
  confirmPane: boolean,
  selectedAsset: string,
}

type State = {
  sendAmount: string,
  sendAddress: string,
}

class Send extends Component<Props, State> {
  state = {
    sendAmount: '',
    sendAddress: ''
  }

  confirmButton: ?HTMLButtonElement

  validateForm () {
    const { dispatch, neo, gas, selectedAsset } = this.props
    const { sendAmount, sendAddress } = this.state
    const neoBalance = neo
    const gasBalance = gas

    if (!sendAddress || !sendAmount) { return false }

    const dispatchClearTransactionEvent = () => setTimeout(() => dispatch(clearTransactionEvent()), 5000)

    try {
      if (verifyAddress(sendAddress) !== true || sendAddress.charAt(0) !== 'A') {
        dispatch(sendEvent(false, 'The address you entered was not valid.'))
        dispatchClearTransactionEvent()
        return false
      }
    } catch (e) {
      dispatch(sendEvent(false, 'The address you entered was not valid.'))
      dispatchClearTransactionEvent()
      return false
    }
    if (selectedAsset === ASSETS_LABELS.NEO && parseFloat(sendAmount) !== parseInt(sendAmount)) { // check for fractional neo
      dispatch(sendEvent(false, 'You cannot send fractional amounts of Neo.'))
      dispatchClearTransactionEvent()
      return false
    } else if (selectedAsset === ASSETS_LABELS.NEO && parseInt(sendAmount) > neoBalance) { // check for value greater than account balance
      dispatch(sendEvent(false, 'You do not have enough NEO to send.'))
      dispatchClearTransactionEvent()
      return false
    } else if (selectedAsset === ASSETS_LABELS.GAS && parseFloat(sendAmount) > gasBalance) {
      dispatch(sendEvent(false, 'You do not have enough GAS to send.'))
      dispatchClearTransactionEvent()
      return false
    } else if (parseFloat(sendAmount) < 0) { // check for negative asset
      dispatch(sendEvent(false, 'You cannot send negative amounts of an asset.'))
      dispatchClearTransactionEvent()
      return false
    }
    return true
  }

  // open confirm pane and validate fields
  openAndValidate = () => {
    const { dispatch } = this.props
    if (this.validateForm()) {
      dispatch(togglePane('confirmPane'))
    }
  }

  // perform send transaction
  sendTransaction = () => {
    const { dispatch, net, address, wif, selectedAsset } = this.props
    const { sendAddress, sendAmount } = this.state

    let assetName
    if (selectedAsset === ASSETS_LABELS.NEO) {
      assetName = ASSETS.NEO
    } else if (selectedAsset === ASSETS_LABELS.GAS) {
      assetName = ASSETS.GAS
    } else {
      dispatch(sendEvent(false, 'That asset is not Neo or Gas'))
      setTimeout(() => dispatch(clearTransactionEvent()), 5000)
      return false
    }
    // validate fields again for good measure (might have changed?)
    if (this.validateForm()) {
      const selfAddress = address
      let sendAsset = {}
      sendAsset[assetName] = sendAmount
      dispatch(sendEvent(true, 'Processing...'))
      log(net, 'SEND', selfAddress, { to: sendAddress, asset: selectedAsset, amount: sendAmount })

      ledgerNanoSGetdoSendAsset(net, sendAddress, wif, sendAsset).then((response) => {
        if (response.result === undefined || response.result === false) {
          dispatch(sendEvent(false, 'Transaction failed!'))
        } else {
          console.log(response.result)
          dispatch(sendEvent(true, 'Transaction complete! Your balance will automatically update when the blockchain has processed it.'))
        }
        setTimeout(() => dispatch(clearTransactionEvent()), 5000)
      }).catch((e) => {
        dispatch(sendEvent(false, 'Transaction failed!'))
        setTimeout(() => dispatch(clearTransactionEvent()), 5000)
      })
      if (this.confirmButton) {
        this.confirmButton.blur()
      }
    }
    // close confirm pane and clear fields
    dispatch(togglePane('confirmPane'))
    this.resetForm()
  }

  resetForm () {
    this.setState({
      sendAddress: '',
      sendAmount: ''
    })
  }

  render () {
    const { dispatch, confirmPane, selectedAsset } = this.props
    const { sendAddress, sendAmount } = this.state
    const confirmPaneClosed = confirmPane ? '100%' : '69%'

    return (
      <SplitPane className='confirmSplit' split='horizontal' size={confirmPaneClosed} allowResize={false}>
        <div id='sendPane'>
          <div id='sendAddress'>
            <input
              type='text'
              placeholder='Where to send the asset (address)'
              value={sendAddress}
              onChange={(e) => this.setState({ sendAddress: e.target.value })}
            />
          </div>
          <div id='sendAmount'>
            <input
              type='text'
              value={sendAmount}
              placeholder='Amount'
              onChange={(e) => this.setState({ sendAmount: e.target.value })}
            />
          </div>
          <button id='sendAsset' data-tip data-for='assetTip' onClick={() => dispatch(toggleAsset())}>{selectedAsset}</button>
          <ReactTooltip class='solidTip' id='assetTip' place='bottom' type='dark' effect='solid'>
            <span>Toggle NEO / GAS</span>
          </ReactTooltip>
          <button id='doSend' onClick={this.openAndValidate}>Send Asset</button>
        </div>
        <div id='confirmPane' onClick={this.sendTransaction}>
          <button ref={node => { this.confirmButton = node }}>Confirm Transaction</button>
        </div>
      </SplitPane>
    )
  }
}

const mapStateToProps = (state) => ({
  wif: state.account.wif,
  address: state.account.address,
  net: state.metadata.network,
  neo: state.wallet.Neo,
  gas: state.wallet.Gas,
  selectedAsset: state.transactions.selectedAsset,
  confirmPane: state.dashboard.confirmPane
})

export default connect(mapStateToProps)(Send)
