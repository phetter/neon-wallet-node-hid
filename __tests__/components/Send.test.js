import React from 'react'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import { shallow, mount } from 'enzyme'
import { TOGGLE_SEND_PANE } from '../../app/modules/dashboard'
import { CLEAR_TRANSACTION, SEND_TRANSACTION, TOGGLE_ASSET } from '../../app/modules/transactions'
import Send from '../../app/components/Send'
import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'

jest.useFakeTimers()
jest.mock('neon-js')

const axiosMock = new MockAdapter(axios)
axiosMock.onAny().reply(200)

const initialState = {
  account: {
    wif: 'L4SLRcPgqNMAMwM3nFSxnh36f1v5omjPg3Ewy1tg2PnEon8AcHou',
    address: 'AWy7RNBVr9vDadRMK9p7i7Z1tL7GrLAxoh'
  },
  metadata: {
    network: 'TestNet'
  },
  wallet: {
    Neo: 5,
    Gas: 1
  },
  transactions: {
    selectedAsset: 'Neo'
  },
  dashboard: {
    confirmPane: true
  }
}

const setup = (state = initialState, shallowRender = true) => {
  const store = configureStore()(state)

  let wrapper
  if (shallowRender) {
    wrapper = shallow(<Send store={store} />)
  } else {
    wrapper = mount(
      <Provider store={store}>
        <Send />
      </Provider>
    )
  }

  return {
    store,
    wrapper
  }
}

const changeAmount = (amount, wrapper) => {
  const amountField = wrapper.find('input[placeholder="Amount"]')
  amountField.instance().value = amount
  amountField.simulate('change')
}

const changeAddress = (address, wrapper) => {
  const addressField = wrapper.find('input[placeholder="Where to send the asset (address)"]')
  addressField.instance().value = address
  addressField.simulate('change')
}

describe('Send', () => {
  test('renders without crashing', (done) => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot()
    done()
  })

  test('toggleAsset button is getting called on click', (done) => {
    const { wrapper, store } = setup()
    const deepWrapper = wrapper.dive()

    deepWrapper.find('#sendAsset').simulate('click')

    const actions = store.getActions()
    expect(actions.length === 1).toEqual(true)
    expect(actions[0].type === TOGGLE_ASSET).toEqual(true)
    done()
  })

  test('sendAsset button is getting called correctly with no address or amount sent', (done) => {
    const { wrapper, store } = setup(initialState, false)
    wrapper.find('#doSend').simulate('click')

    jest.runAllTimers()
    const actions = store.getActions()
    expect(actions.length === 0).toEqual(true)
    done()
  })

  test('sendAsset button is getting called correctly with invalid address', (done) => {
    const { wrapper, store } = setup(initialState, false)

    changeAddress('This-is-an-invalid-address', wrapper)
    changeAmount(initialState.wallet.Neo, wrapper)

    wrapper.find('#doSend').simulate('click')

    jest.runAllTimers()
    const actions = store.getActions()
    expect(actions.length === 2).toEqual(true)
    expect(actions[0]).toEqual({
      type: SEND_TRANSACTION,
      success: false,
      message: 'The address you entered was not valid.'
    })
    expect(actions[1]).toEqual({
      type: CLEAR_TRANSACTION
    })
    done()
  })

  test('sendAsset button is getting called correctly with fractional NEO', (done) => {
    const { wrapper, store } = setup(initialState, false)

    changeAddress(initialState.account.address, wrapper)
    changeAmount(initialState.wallet.Neo + 0.5, wrapper)

    wrapper.find('#doSend').simulate('click')

    jest.runAllTimers()
    const actions = store.getActions()
    expect(actions.length === 2).toEqual(true)
    expect(actions[0]).toEqual({
      type: SEND_TRANSACTION,
      success: false,
      message: 'You cannot send fractional amounts of Neo.'
    })
    expect(actions[1]).toEqual({
      type: CLEAR_TRANSACTION
    })
    done()
  })

  test('sendAsset button is getting called correctly with no NEO amount', (done) => {
    const { wrapper, store } = setup(initialState, false)

    changeAddress(initialState.account.address, wrapper)

    wrapper.find('#doSend').simulate('click')

    jest.runAllTimers()
    const actions = store.getActions()
    expect(actions.length === 0).toEqual(true)
    done()
  })

  test('sendAsset button is getting called correctly with not enough NEO', (done) => {
    const { wrapper, store } = setup(initialState, false)

    changeAddress(initialState.account.address, wrapper)
    changeAmount(initialState.wallet.Neo + 1, wrapper)

    wrapper.find('#doSend').simulate('click')

    jest.runAllTimers()
    const actions = store.getActions()
    expect(actions.length === 2).toEqual(true)
    expect(actions[0]).toEqual({
      type: SEND_TRANSACTION,
      success: false,
      message: 'You do not have enough NEO to send.'
    })
    expect(actions[1]).toEqual({
      type: CLEAR_TRANSACTION
    })
    done()
  })

  test('sendAsset button is getting called correctly with negative NEO', (done) => {
    const { wrapper, store } = setup(initialState, false)

    changeAddress(initialState.account.address, wrapper)
    changeAmount(-1, wrapper)

    wrapper.find('#doSend').simulate('click')

    jest.runAllTimers()
    const actions = store.getActions()
    expect(actions.length === 2).toEqual(true)
    expect(actions[0]).toEqual({
      type: SEND_TRANSACTION,
      success: false,
      message: 'You cannot send negative amounts of an asset.'
    })
    expect(actions[1]).toEqual({
      type: CLEAR_TRANSACTION
    })
    done()
  })

  test('sendAsset button is getting called correctly with correct NEO amount', (done) => {
    const { wrapper, store } = setup(initialState, false)

    changeAddress(initialState.account.address, wrapper)
    changeAmount(initialState.wallet.Neo - 1, wrapper)

    wrapper.find('#doSend').simulate('click')

    const actions = store.getActions()
    expect(actions.length === 1).toEqual(true)
    expect(actions[0]).toEqual({
      type: TOGGLE_SEND_PANE,
      pane: 'confirmPane'
    })
    done()
  })

  test('sendAsset button is getting called correctly for without enough Gas', (done) => {
    const gasState = Object.assign({}, initialState, { transactions: { selectedAsset: 'Gas' } })
    const { wrapper, store } = setup(gasState, false)

    changeAddress(initialState.account.address, wrapper)
    changeAmount(initialState.wallet.Gas + 1, wrapper)

    wrapper.find('#doSend').simulate('click')

    jest.runAllTimers()
    const actions = store.getActions()
    expect(actions.length === 2).toEqual(true)
    expect(actions[0]).toEqual({
      type: SEND_TRANSACTION,
      success: false,
      message: 'You do not have enough GAS to send.'
    })
    expect(actions[1]).toEqual({
      type: CLEAR_TRANSACTION
    })
    done()
  })

  test('sendAsset button is getting called correctly with correct Gas amount', (done) => {
    const gasState = Object.assign({}, initialState, { transactions: { selectedAsset: 'Gas' } })
    const { wrapper, store } = setup(gasState, false)

    changeAddress(initialState.account.address, wrapper)
    changeAmount(initialState.wallet.Gas - 0.5, wrapper)

    wrapper.find('#doSend').simulate('click')

    const actions1 = store.getActions()
    expect(actions1.length === 1).toEqual(true)
    expect(actions1[0]).toEqual({
      type: TOGGLE_SEND_PANE,
      pane: 'confirmPane'
    })
    done()
  })

  test('send transaction is working correctly', (done) => {
    const { wrapper, store } = setup(initialState, false)

    changeAddress(initialState.account.address, wrapper)
    changeAmount(initialState.wallet.Neo - 1, wrapper)

    wrapper.find('#confirmPane').simulate('click')

    Promise.resolve('pause').then(() => {
      jest.runAllTimers()
      const actions = store.getActions()
      expect(actions.length === 4).toEqual(true)
      expect(actions[0]).toEqual({
        type: SEND_TRANSACTION,
        success: true,
        message: 'Processing...'
      })
      expect(actions[1]).toEqual({
        type: TOGGLE_SEND_PANE,
        pane: 'confirmPane'
      })
      expect(actions[2]).toEqual({
        type: SEND_TRANSACTION,
        success: true,
        message: 'Transaction complete! Your balance will automatically update when the blockchain has processed it.'
      })
      expect(actions[3]).toEqual({
        type: CLEAR_TRANSACTION
      })
      done()
    }).catch(e => done.fail(e))
  })

  test('component is rendering correctly', (done) => {
    const { wrapper } = setup()
    const deepWrapper = wrapper.dive()

    const checkElements = [
      {
        element: deepWrapper.find('#sendAsset'),
        text: 'Neo'
      },
      {
        element: deepWrapper.find('#sendAmount'),
        text: ''
      },
      {
        element: deepWrapper.find('#sendAddress'),
        text: ''
      },
      {
        element: deepWrapper.find('#doSend'),
        text: 'Send Asset'
      },
      {
        element: deepWrapper.find('#confirmPane'),
        text: 'Confirm Transaction'
      }
    ]

    checkElements.forEach(item => {
      expect(item.element.length).toEqual(1)
      expect(item.element.text()).toEqual(item.text)
    })
    done()
  })
})
