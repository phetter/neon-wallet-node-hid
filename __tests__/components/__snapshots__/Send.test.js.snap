// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Send renders without crashing 1`] = `
<Send
  address="AWy7RNBVr9vDadRMK9p7i7Z1tL7GrLAxoh"
  confirmPane={true}
  dispatch={[Function]}
  gas={1}
  neo={5}
  net="TestNet"
  selectedAsset="Neo"
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  wif="L4SLRcPgqNMAMwM3nFSxnh36f1v5omjPg3Ewy1tg2PnEon8AcHou"
/>
`;
